Add execution stack design doc. (#127)
BranchTaken/Hemlock: doc/design/execution_stack.md

```hemlock
older_pc_address = current_pc_address + (4 * current_frame_size)
```
#### C frames
<!--
- If a PC falls outside of a Hemlock code segment, we determine that it is part of a C frame.
- Use DWARF info to progressively unroll until we hit a Hemlock PC again.
- Store the combined size of the C frames
-->
#### Hemlock reentry from C
<!--
- Use libunwind to get stack info. Store the size of the whole blob in a table keyed by the SP where
  the PC is.
- XXX How we deal with VLA in a C frame.
- In the reentry Hemlock frame, store a value (initially 0) for the size of the combined C frames
  above. Upon unrolling the stack, if this value is 0, calculate the size of the C frames and then
  store the value here.
-->

### Garbage collection

Stack frames include both (GC-able) references and (non-GC-able) non-references. The garbage
collector must know where references exist within each stack frame so that it may update them during
[minor heap copy-collection](memory.md#minor-heap) and [major heap
compaction](memory.md#major-heap). The PC metadata table includes a bitmap of the entire stack frame
that marks each address as either a reference (1) or a non-reference (0).

```hemlock

> let gc_frame address frame_metadata gc_state =
>   let rec inner address bitmap gc_state =
>     match bitmap = 0 with
>     | true -> gc_state
>     | false ->
>       n_zeros = clz bitmap
>       address = address + (8 * n_zeros)
>       GarbageCollection.collect address gc_state
>        |> inner address (bitmap >> (n_zeros + 1))
>   inner address frame_metadata.gc_references_bitmap gc_state
> 
> let rec gc_stack pc_address gc_state =
>   match pc_address >= gc_state.stack_address with
>   | true -> gc_state
>   | false ->
>     let frame_metadata = FrameMetadata.lookup pc_address.value
>     gc_frame pc_address frame_metadata.gc_references_bitmap gc_state
>      |> gc_stack pc_address + (8 * frame_metadata.frame_size)

Reviewer: ```suggestion
let gc_frame address frame_metadata gc_state =
    let rec inner address bitmap gc_state =
      match Bitmap.(bitmap = 0) with
      | true -> gc_state
      | false ->
          let n_zeros = Bitmap.ctz bitmap
          let address' = address + (word_size * n_zeros)
          let bitmap' = Bitmap.usr (n_zeros + 1) bitmap
          GarbageCollection.mark address' gc_state
            |> inner address' bitmap'
    inner address frame_metadata.gc_references_bitmap gc_state

let rec gc_stack pc_address gc_state =
    match pc_address >= gc_state.stack_address with
    | true -> gc_state
    | false ->
        let frame_metadata = FrameMetadata.lookup pc_address.value
        gc_frame pc_address frame_metadata.gc_references_bitmap gc_state
          |> gc_stack (pc_address + (word_size * frame_metadata.frame_size))
```

```

### Work stealing

> Some stack frame include set-up for a single stealable [right

Reviewer: ```suggestion
Some stack frames include setup for a single stealable [right
```

continuation](work_stealing.md#terminology) (RC). If there is a stealable RC in a stack frame, the

> presence is marked via a non-zero (quantum generation)[quantum_generation] offset value in the stack frame's metadata. An

Reviewer: ```suggestion
presence is marked via a non-zero (quantum generation)[quantum_generation] (QG) offset value in the stack frame's metadata. An
```

offset value of `0` from the PC address would not make sense, so is reserved to indicate that no
RC can be stolen from the stack frame.

```hemlock

> let ws_frame pc_address frame_metadata ws_state =
>   match frame_metadata.quantum_generation_offset = 0 with
>   | true -> ws_state  // No work stealing opportunities in this frame
>   | false ->
>     quantum_generation_address = pc_address + (8 * frame_metadata.quantum_generation_offset)
>     match quantum_generation_address.value = 0 with
>     | true -> WorkStealingState.set_quantum_generation quantum_generation_address ws_state
>     | false ->
>       match quantum_generation_address.value > ws_state.stealable_quantum_generation with
>       | true -> ws_state  // This frame hasn't survived on the stack long enough for work stealing
>       | false -> WorkStealing.steal pc_address frame_metadata.ws_trampoline_pc ws_state
> 
> let rec ws_stack pc_address ws_state =
>   match_pc_address >= ws_state.stack_address with
>   | true -> ws_state
>   | false ->
>     let frame_metadata = FrameMetadata.lookup pc_address.value
>     ws_frame pc_address frame_metadata ws_state
>      |> ws_stack pc_address + (8 * frame_metadata.frame_size)

Reviewer: ```suggestion
let ws_frame pc_address frame_metadata ws_state =
    match frame_metadata.quantum_generation_offset = 0 with
    | true -> ws_state # No work stealing opportunities in this frame
    | false ->
        quantum_generation_address = pc_address + (word_size * frame_metadata.quantum_generation_offset)
        match quantum_generation_address.value = 0 with
        | true -> WorkStealingState.set_quantum_generation quantum_generation_address ws_state
        | false ->
            match quantum_generation_address.value > ws_state.stealable_quantum_generation with
            | true -> ws_state # This frame hasn't survived on the stack long enough for work stealing
            | false -> WorkStealing.steal pc_address frame_metadata.ws_trampoline_pc ws_state

let rec ws_stack pc_address ws_state =
    match pc_address >= ws_state.stack_address with
    | true -> ws_state
    | false ->
        let frame_metadata = FrameMetadata.lookup pc_address.value
        ws_frame pc_address frame_metadata ws_state
          |> ws_stack pc_address + (word_size * frame_metadata.frame_size)
```

```
#### Trampoline function

If an RC can be stolen from the stack frame, stack frame metadata will include a valid
`ws_trampoline_pc` field. The work stealing trampoline PC points to a generated bespoke trampoline
function for launching the RC in a (delegate)[work_stealing.md#terminology] actor. Given the PC
address of the (victim)[work_stealing.md#terminology] actor's stack frame, the bespoke trampoline
function does the following:

1. Rewrites the victim stack frame's PC to point to a merge function generated specifically for the

>    (left continuation (LC))[work_stealing.md#terminology] to merge its result with the

Reviewer: ```suggestion
   [left continuation (LC)](work_stealing.md#terminology) to merge its result with the
```

   already-present result of the RC (see final trampoline function step).
2. Loads the victim's `RC in-register args...` section of the stack frame into registers.
3. Copies the victim's `RC on-stack args...` to the delegate's stack and calls the RC.
4. Returns the result of the RC into the victim's stack frame. The entire space in the victim stack
   frame's `Quantum generation`, `RC on-stack-args...`, and `RC in-register args...` sections is now
   available for copying a result of constant size. If the result value is of variable size, is a

>    boxed value, or is simply a constant size simply too large to fit in the available space, the

Reviewer: ```suggestion
   boxed value, or is simply a constant size too large to fit in the available space, the
```

>    value is boxed. The box is written to the stack with the value is written to the victim actor's
>    heap.

Reviewer: ```suggestion
   value is boxed. If the result would be a reference to a value in the delegate's heap, a copy of the result is written to a
   location with lifetime longer than the delegate's heap (global heap, victim's heap/mailbox, etc.); mechanism TBD.
```
